kin-devplatform-sdk/src/main/java/kin/devplatform/KinEcosystemInitiator.java

```diff
@@ -76,85 +80,109 @@ public void internalInit(Context context) {
-		final KinCallback<Void> loginCallback) {
+	public void externalInit(Context context, String appId, KinEnvironment environment, @NonNull SignInData signInData,
+							 final KinCallback<Void> loginCallback) {
 		if (isInitialized && isLoggedIn) {
 			fireStartCompleted(loginCallback);
 			return;
 		}
 
 		try {
-			init(context, environment);
+			init(context, appId, environment, true, signInData, loginCallback);
 		} catch (BlockchainException e) {
 			fireStartError(e, loginCallback);
-			return;
 		}
-		login(signInData, loginCallback);
 	}
 
-	private void init(Context context, KinEnvironment environment) throws BlockchainException {
+	private void init(Context context, String appId, KinEnvironment environment, boolean withLogin,
+					  SignInData signInData, KinCallback<Void> loginCallback) throws BlockchainException {
 		if (!isInitialized) {
 			ConfigurationLocal configurationLocal = ConfigurationLocal.getInstance(context);
 			if (environment != null) {
 				configurationLocal.setEnvironment(environment);
 			}
+
 			ConfigurationImpl.init(configurationLocal);
 			KinEnvironment kinEnvironment = ConfigurationImpl.getInstance().getEnvironment();
-			EventLogger eventLogger = EventLoggerImpl.getInstance();
-			final String networkUrl = kinEnvironment.getBlockchainNetworkUrl();
-			final String networkId = kinEnvironment.getBlockchainPassphrase();
+			final EventLogger eventLogger = EventLoggerImpl.getInstance();
+			final String oldNetworkUrl = kinEnvironment.getOldBlockchainNetworkUrl();
+			final String oldNetworkId = kinEnvironment.getOldBlockchainPassphrase();
+			final String newNetworkUrl = kinEnvironment.getNewBlockchainNetworkUrl();
+			final String newNetworkId = kinEnvironment.getNewBlockchainPassphrase();
 			final String issuer = kinEnvironment.getIssuer();
 
-			KinClient kinClient = new KinClient(context, new ServiceProvider(networkUrl, networkId) {
-				@Override
-				protected String getIssuerAccountId() {
-					return issuer;
-				}
-			}, KIN_ECOSYSTEM_STORE_PREFIX_KEY);
-			BlockchainSourceImpl.init(eventLogger, kinClient, BlockchainSourceLocal.getInstance(context));
+			// TODO: 01/01/2019 is it ok to use issuer like i did or like they did in the past like this:
+//			new ServiceProvider(networkUrl, networkId) {
+//				@Override
+//				protected String getIssuerAccountId() {
+//					return issuer;
+//				}
 
-			AuthRepository
-				.init(AuthLocalData.getInstance(context, executorsUtil),
-					AuthRemoteData.getInstance(executorsUtil));
+			MigrationNetworkInfo migrationNetworkInfo = new MigrationNetworkInfo(oldNetworkUrl, oldNetworkId, newNetworkUrl, newNetworkId, issuer);
+			MigrationManager migrationManager = new MigrationManager(context, appId, migrationNetworkInfo, new KinVersionProvider(), KIN_ECOSYSTEM_STORE_PREFIX_KEY);
+			try {
+				handleMigration(context, appId, eventLogger, migrationManager, withLogin, signInData, loginCallback);
+			} catch (MigrationInProcessException e) {
```

Comment:
It is a special case where the developers call start while it is already in migration process.
The start method declare that it can throw this exception in case they call the method while the method didn't finished yet(onError or onReady called).
It seems like other APIs are doing such stuff, meaning if the method is asynchronous and some error will occur then it will call onError but in this specific case the asynchronous code is not even called yet because it is immediately(on the same called thread) will throw this error if needed.

Anyway if you really think it is a good practice also in this case to use the onError callback then I can change it easily.

Reply:
@yosriz ,so what do you think?